File: leetcode_alg/solution199.cpp
//
// Created by wangheng on 2022/6/18.
//

#include <iostream>
#include <vector>
#include <queue>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<int> rightSideView(TreeNode *root) {
        vector<int> res;
        if (root == nullptr) {
            return res;
        }
        queue<TreeNode *> q;
        q.push(root);
        while (!q.empty()) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                TreeNode *tmpNode = q.front();
                q.pop();
                if (tmpNode->left != nullptr) {
                    q.push(tmpNode->left);
                }
                if (tmpNode->right != nullptr) {
                    q.push(tmpNode->right);
                }
                if (i == size - 1) {
                    res.push_back(tmpNode->val);
                }
            }
        }
        return res;
    }
};

int main() {
    Solution *s = new Solution;
    TreeNode *root = new TreeNode(1, new TreeNode(2, nullptr, new TreeNode(5)),
                                  new TreeNode(3, nullptr, new TreeNode(4)));
    vector<int> res = s->rightSideView(root);
    for (int a: res) {
        cout << a << " ";
    }
    cout << endl;
    return 0;
}

File: golang/solution/solution199.go
package solution

import "fmt"

// 二叉树的右视图

// 方法1 层序遍历
func rightSideView1(root *TreeNode) []int {
	var res []int = make([]int, 0)
	if root == nil {
		return res
	}
	var queue []*TreeNode = make([]*TreeNode, 0)
	queue = append(queue, root)
	for len(queue) > 0 {
		var size int = len(queue)
		for i := 0; i < size; i++ {
			var node *TreeNode = queue[0]
			queue = queue[1:]
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
			if i == size-1 {
				res = append(res, node.Val)
			}
		}
	}
	return res
}

func Solution199() {

	var root *TreeNode = &TreeNode{
		Val:   1,
		Left:  &TreeNode{Val: 2, Left: &TreeNode{Val: 4}, Right: &TreeNode{Val: 5}},
		Right: &TreeNode{Val: 3},
	}

	res := rightSideView1(root)
	fmt.Println(res)
}
